LFS-823: As a user, I am stuck in a relogin cycle when deleting a form while editing that form

IntegralProgrammer opened the pull request:
Fixes the bug present on the `dev` branch described by LFS-823

To reproduce that bug on the `dev` branch:

1. Create any form
2. While in the form, click the trash can icon to delete the form and confirm
3. You are prompted to login, logging in reopens the same login screen - this continues without end
4. To get out of the loop, you must clear the URL and return home

To test this PR:

1. Build the `LFS-823` branch
2. Create a new _Demographics_ form, enter some data, and close that browser tab without clicking the _Save_ button.
3. Open CARDS again and this _Demographics_ form should be present with the data entered from the previous step
4. Enter some more data in this same _Demographics_ form and click on one of the sidebar links (don't click _Save_)
5. Open this _Demographics_ form again - the data from the previous step should be present
6. Delete this form - you should not be prompted to re-login

IntegralProgrammer committed:
LFS-823: As a user, I am stuck in a relogin cycle when deleting a form while editing that form

Disconnect the beforeunload event listener (which automatically saves the Form) when the Delete button is clicked

Send the beforeunload event whenever the URL in the browser's address bar changes - such as when clicking a link on the sidebar.

marta- commented on modules/data-entry/src/main/frontend/src/questionnaire/Form.jsx:
```diff
@@ -319,6 +323,7 @@ function Form (props) {
               entryType={data?.questionnaire?.title || "Form"}
               shouldGoBack={true}
               buttonClass={classes.titleButton}
+              removeWindowHandlers={removeWindowHandlers}
```
I don't think we need to be so specific and add a property named `removeWindowHandlers` to the `DeleteButton`. Technically this is something (anything) that is supposed to happen once _deletion is successful_. So we can make use of the `onComplete` handler and get rid of all changes done to the `DeleteButton` component.
```suggestion
              onComplete={removeWindowHandlers}
```

IntegralProgrammer replied:
Implemented by commit `8d67b37` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -29,7 +29,7 @@ import QuestionnaireStyle from "../questionnaire/QuestionnaireStyle.jsx";
  */
 function DeleteButton(props) {
-  const { classes, entryPath, entryName, onComplete, entryType, entryLabel, size, shouldGoBack, buttonClass } = props;
+  const { classes, entryPath, entryName, onComplete, entryType, entryLabel, size, shouldGoBack, buttonClass, removeWindowHandlers } = props;
```
```suggestion
  const { classes, entryPath, entryName, onComplete, entryType, entryLabel, size, shouldGoBack, buttonClass} = props;
```

IntegralProgrammer replied:
Implemented by commit `8d67b37` :heavy_check_mark:

marta- commented on modules/data-entry/src/main/frontend/src/dataHomepage/DeleteButton.jsx:
```diff
@@ -137,6 +137,7 @@ function DeleteButton(props) {
     if (deleteRecursive) {
       url.searchParams.set("recursive", true);
     }
+    if (removeWindowHandlers) {removeWindowHandlers();}
```
Handlers should be removed only when deletion succeeded and we actually need to navigate away.
```suggestion
```

IntegralProgrammer replied:
Implemented by commit `8d67b37` :heavy_check_mark:

marta- reviewed (changes requested):
Works fine except that handlers are removed even if the deletion doesn't go through.

IntegralProgrammer committed:
LFS-823: As a user, I am stuck in a relogin cycle when deleting a form while editing that form

Handle the removal of the beforeunload event listener when deleting a form through the onComplete prop of the DeleteButton component.

Remove the removeWindowHandlers prop of the DeleteButton as it is no longer needed.

IntegralProgrammer committed:
LFS-823: As a user, I am stuck in a relogin cycle when deleting a form while editing that form

Rename removeThisHandler to removeSaveDataHandler

IntegralProgrammer committed:
LFS-823: As a user, I am stuck in a relogin cycle when deleting a form while editing that form

Fix bug where JSON Form contents was unnecessarily fetched twice

marta- reviewed (approved):
Looks great

sashaandjic reviewed (approved):
tested

sashaandjic commented:
Please note that sometimes I get prompted to login after these steps - however I cannot always reproduce it; it happens once in 6-7 attempts:
- create new Demographic form
- create new subject for the form
- edit the form and close the browser tab without saving the form
- open CARDS - sometimes (rarely) I get prompted to login, after which no data is lost and everything works as expected
